### DISCUSSION TITLE
Readability vs Optimization

### DISCUSSION THREAD

**Discussion by edhans** [tabulartools/dax-optimizer#3] [Q&A]
Just throwing this out there to see what opinions are.

I had the following in a measure. DAX Optimizer said I had duplicated the VALUES ( 'Dataflow Status'[Snapshot Date] ) several times (two in this example) So I replaced it with Version 2 below. Both here return BLANK() just so DAX Formatter would work.
```
Version 1 =
VAR varSelectedSnapshot =
    SELECTEDVALUE ( 'Dataflow Status'[Snapshot Type] )
VAR varFirstFactTable =
    CALCULATE (
        SUM ( 'Fact Table 1'[Quantity] ),
        TREATAS (
            VALUES ( 'Dataflow Status'[Snapshot Date] ),
            'Fact Table 1'[Snapshot Date]
        ),
        REMOVEFILTERS ( 'Dim Table 1' ),
        REMOVEFILTERS ( 'Dim Table 2' ),
        REMOVEFILTERS ( 'Dim Table 3' )
    )
VAR varSecondFactTable =
    CALCULATE (
        SUM ( 'Fact Table 2'[Quantity] ),
        TREATAS (
            VALUES ( 'Dataflow Status'[Snapshot Date] ),
            'Fact Table 2'[Snapshot Date]
        ),
        REMOVEFILTERS ( 'Dim Table 1' ),
        REMOVEFILTERS ( 'Dim Table 2' ),
        REMOVEFILTERS ( 'Dim Table 3' )
    )
RETURN
    BLANK ()
```
Now with VALUES ( 'Dataflow Status'[Snapshot Date] ) in a variable.

```
Version 2 =
VAR varSelectedSnapshot =
    SELECTEDVALUE ( 'Dataflow Status'[Snapshot Type] )
VAR varSnapshotsVisible =
    VALUES ( 'Dataflow Status'[Snapshot Date] )
VAR varFirstFactTable =
    CALCULATE (
        SUM ( 'Fact Table 1'[Quantity] ),
        TREATAS (
            varSnapshotsVisible,
            'Fact Table 1'[Snapshot Date]
        ),
        REMOVEFILTERS ( 'Dim Table 1' ),
        REMOVEFILTERS ( 'Dim Table 2' ),
        REMOVEFILTERS ( 'Dim Table 3' )
    )
VAR varSecondFactTable =
    CALCULATE (
        SUM ( 'Fact Table 2'[Quantity] ),
        TREATAS (
            varSnapshotsVisible,
            'Fact Table 2'[Snapshot Date]
        ),
        REMOVEFILTERS ( 'Dim Table 1' ),
        REMOVEFILTERS ( 'Dim Table 2' ),
        REMOVEFILTERS ( 'Dim Table 3' )
    )
RETURN
    BLANK ()
```

That is way harder to read IMHO. I always expect a table as the first parameter of TREATAS vs having a variable that is a table.

From a performance standpoint on _this example_ there is no material difference - this is a diagnostic table of about 50 records max that I am getting VALUES from and there are no relationships with it.

The physical query plans were identical. The only differences I could see in the logical query plans were:

1. An extra line to assign VALUES() to the varSnapshotsVisible variable
2. In the query plan where I left it as tables with no variable, there was an extra Constant: ScaLogOp DependOnCols line and ScalarVarProxy line

Neither of which I will pretend to understand.

So ignoring the size of the table - say my diag table was a 800K record products table, if the physical plan would be the same anyway, would it make a difference? And if not, should this be taken into account when the same table is repeated multiple times in a series of TREATAS() functions like this?

**Comment by marcosqlbi** [comment]
I understand your point. However, we decided getting a false positive (which you can ignore) rather than a false negative (which you wouldn't see) is better. Therefore, in a case like this one, it is true that there are no differences, but we know that in more complex measures, there is an advantage in storing a table expression in a variable. We are not 100% sure that VALUES ( column ) would be able to be optimized by the query plan in any complex conditions (we tend to say "no" because of past experiences, even though we do not have a simple repro to show). 
I would keep the conversation open to get additional feedback. Longer term, we would like to have attributes/comments to tell DAX Optimizer that a certain coding choice should ignore a warning, but it's not planned for the short term. But we could always change prioritization if needed.
Thanks!
